### PR TITLE
#4452 [Manager] Refactor connector API to use the new architecture

### PR DESCRIPTION
**Pull request by wu87988622**
### Checklist
- [x] [Review Contribution Guidelines](https://ohara.readthedocs.io/en/latest/contributing.html)
- [x] Add Reviewers (see [Authors](https://github.com/oharastream/ohara#ohara-team))
- [x] Add Assignees (of course, you are the assignee by default!)
- [x] DON'T set a Milestone
- [x] DON'T add any labels (PR labels will be added automatically by Jenkins)
- [x] Link to an issue(You can do so by using the required by keyword: required by #4452 )
- [ ] Add unit tests (or please explain why this PR is not worth having new tests)
- [ ] Pass QA (Sometimes you may encounter an existing flaky test which would obstruct you from getting QA passed. If the failed tests are unrelated, you can add a comment in the PR thread)

### CODE REVIEW

File: ohara-manager/client/src/store/epics/connector/removeSinkLinkEpic.js [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+  return action$.pipe(
+    ofType(actions.removeConnectorSinkLink.TRIGGER),
+    map(action => action.payload),
+    switchMap(values =>
+      from(connectorApi.update(values)).pipe(
+        map(res => normalize(res.data, schema.connector)),
+        map(entities => actions.removeConnectorSinkLink.success(entities)),
+        startWith(actions.removeConnectorSinkLink.request()),
+        catchError(err => {
+          values.paperApi.addLink(values.id, values.topic.id);
```
Comment (saivirtue) [inline ohara-manager/client/src/store/epics/connector/removeSinkLinkEpic.js:36]: should be `removeLink`?

File: ohara-manager/client/src/hooks/connectorHooks.js [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+  const dispatch = useDispatch();
+  const group = useConnectorGroup();
+  return values =>
+    dispatch(
+      actions.removeConnectorSinkLink.trigger({
+        ...values,
+        group,
+      }),
+    );
+};
```
Comment (saivirtue) [inline ohara-manager/client/src/hooks/connectorHooks.js:83]: Could we merge these two `useRemoveXXX` into one action?

File: ohara-manager/client/src/hooks/connectorHooks.js [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+  const dispatch = useDispatch();
+  const group = useConnectorGroup();
+  return values =>
+    dispatch(
+      actions.updateConnectorLink.trigger({
+        ...values,
+        group,
+      }),
+    );
+};
```
Comment (saivirtue) [inline ohara-manager/client/src/hooks/connectorHooks.js:59]: 因為在`useUpdateConnectorAction`內也會更新ConnectorLink，我們是否可以將這兩個action合併成一個就好?

File: ohara-manager/client/src/store/epics/connector/removeSinkLinkEpic.js (hunk 1 above)
Comment (wu87988622) [reply ohara-manager/client/src/store/epics/connector/removeSinkLinkEpic.js:36]: 這是如果使用者想刪除某一個Link但是失敗了，我們會將他加回去。

File: ohara-manager/client/src/store/epics/connector/updateLinkConnectorEpic.js [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+import * as schema from 'store/schema';
+
+export default action$ => {
+  return action$.pipe(
+    ofType(actions.updateConnectorLink.TRIGGER),
+    map(action => action.payload),
+    switchMap(values =>
+      from(connectorApi.update(values)).pipe(
+        map(res => normalize(res.data, schema.connector)),
+        map(entities => actions.updateConnectorLink.success(entities)),
```
Comment (eechih) [inline ohara-manager/client/src/store/epics/connector/updateLinkConnectorEpic.js:33]: Please rename `entities` to `normalizedData`

File: ohara-manager/client/src/store/epics/nodeEpics.js [hunk 5]
```diff
@@ -77,13 +77,20 @@ const fetchNodesEpic = action$ =>
...
+      iif(
+        () => res.data.find(node => node.hostname === values),
+        throwError,
+        res,
+      ),
+    ),
+    retryWhen(error =>
+      error.pipe(
+        concatMap((e, i) =>
+          iif(() => i > 4, throwError(e), of(e).pipe(delay(2000))),
```
Comment (eechih) [inline ohara-manager/client/src/store/epics/nodeEpics.js:90]: Good

File: ohara-manager/client/src/store/epics/connector/updateConnectorEpic.js [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+import * as schema from 'store/schema';
+import { KIND } from 'const';
+
+export default action$ => {
+  return action$.pipe(
+    ofType(actions.updateConnector.TRIGGER),
+    map(action => action.payload),
+    switchMap(values => {
+      const hasTopicKey = values.topicKeys.length > 0;
+      const { cell, paperApi, topics } = values;
```
Comment (eechih) [inline ohara-manager/client/src/store/epics/connector/updateConnectorEpic.js:34]: Can we move `cell`, `paperApi` to another parameter? ex: `options`

File: ohara-manager/client/src/store/epics/connector/stopConnectorEpic.js [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+  return action$.pipe(
+    ofType(actions.stopConnector.TRIGGER),
+    map(action => action.payload),
+    switchMap(values => {
+      values.paperApi.updateElement(values.id, {
+        status: CELL_STATUS.pending,
+      });
+      return of(
+        defer(() => connectorApi.stop(values)).pipe(
+          map(() => actions.stopConnector.request()),
```
Comment (eechih) [inline ohara-manager/client/src/store/epics/connector/stopConnectorEpic.js:68]: Why not use `startWith()`?

File: ohara-manager/client/src/hooks/connectorHooks.js (hunk 2 above)
Comment (wu87988622) [reply ohara-manager/client/src/hooks/connectorHooks.js:83]: i will in #4456 add new task

File: ohara-manager/client/src/store/epics/connector/stopConnectorEpic.js [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+import * as actions from 'store/actions';
+import * as schema from 'store/schema';
+import { CELL_STATUS } from 'const';
+
+const checkState$ = values =>
+  defer(() =>
+    connectorApi.get({ name: values.name, group: values.group }),
+  ).pipe(
+    map(res => normalize(res.data, schema.connector)),
+    map(res => iif(() => res?.state, throwError, res)),
```
Comment (eechih) [inline ohara-manager/client/src/store/epics/connector/stopConnectorEpic.js:41]: 這段邏輯有點怪，map 接到的參數應該是 normalize data 而不是 response result.
Please rename `res` to `normalizedData`.

File: ohara-manager/client/src/hooks/connectorHooks.js (hunk 3 above)
Comment (wu87988622) [reply ohara-manager/client/src/hooks/connectorHooks.js:59]: i will in #4456 add new task

File: ohara-manager/client/src/store/epics/connector/stopConnectorEpic.js (hunk 7 above)
Comment (wu87988622) [reply ohara-manager/client/src/store/epics/connector/stopConnectorEpic.js:68]: 這邊是因為需要一個回傳值，of().pip(concatAll())中給一個都必須有回傳值

File: ohara-manager/client/src/store/epics/connector/updateLinkConnectorEpic.js (hunk 4 above)
Comment (wu87988622) [reply ohara-manager/client/src/store/epics/connector/updateLinkConnectorEpic.js:33]: ok

File: ohara-manager/client/src/store/epics/connector/stopConnectorEpic.js (hunk 8 above)
Comment (wu87988622) [reply ohara-manager/client/src/store/epics/connector/stopConnectorEpic.js:41]: ok

File: ohara-manager/client/src/store/epics/connector/updateConnectorEpic.js (hunk 6 above)
Comment (wu87988622) [reply ohara-manager/client/src/store/epics/connector/updateConnectorEpic.js:34]: ok

File: ohara-manager/client/src/store/epics/connector/stopConnectorEpic.js (hunk 8 above)
Comment (eechih) [reply ohara-manager/client/src/store/epics/connector/stopConnectorEpic.js:41]: 補充一下：res?.state 應該不存在，因為 input parameter 是 normalize data 而不是 response result

Comment (wu87988622) [reply ohara-manager/client/src/store/epics/connector/stopConnectorEpic.js:41]: 對!! 所以需要rename，造成誤解

File: ohara-manager/client/src/store/epics/connector/createConnectorEpic.js [hunk 9]
```diff
@@ -28,18 +28,18 @@ export default action$ => {
   return action$.pipe(
     ofType(actions.createConnector.TRIGGER),
     map(action => action.payload),
-    switchMap(values =>
+    switchMap((values, options) =>
```
Comment (eechih) [inline ohara-manager/client/src/store/epics/connector/createConnectorEpic.js:31]: 這樣的寫法能拿到 options 嗎？

🤔 我覺得應該是 `switchMap(({ values, options }) =>`

Comment (wu87988622) [reply ohara-manager/client/src/store/epics/connector/createConnectorEpic.js]: 剛剛有測試出來，我已經調整了
感謝

File: ohara-manager/client/src/components/Pipeline/PipelineApiHelper/connector.js [hunk 10]
```diff
@@ -14,173 +14,76 @@
...
+  const updateConnector = hooks.useUpdateConnectorAction();
+  const updateConnectorLink = hooks.useUpdateConnectorLinkAction();
+  const removeConnectorSourceLink = hooks.useRemoveSourceLinkAction();
+  const removeConnectorSinkLink = hooks.useRemoveSinkLinkAction();
+  const workerClusterKey = hooks.useWorkerClusterKey();
+  const topicGroup = hooks.useTopicGroup();
+
+  const create = async (values, paperApi) => {
+    const { id, name, className } = values;
+    const params = { id, name, connector__class: className, workerClusterKey };
```
Comment (eechih) [inline ohara-manager/client/src/components/Pipeline/PipelineApiHelper/connector.js:33]: 這裡的 id 應該是 paper element id 吧！ 放在 values 好像容易誤會 connector 有一個屬性為 id 。🤔
移到 options 是否比較好呢？另外，是否能給一個較容易辨別的名稱呢？

**Pull request merged**